### PR TITLE
fix: Do not add fake root for searches with context

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
When page source is built for the whole elements hierarchy that a fake `hierarchy` root is added, which contains some additional info about the display. It is also used as a holder for toast elements. Although, it makes no sense to add the same stuff to context searches where the xml root is some subelement.

https://github.com/appium/appium/issues/14509

### CODE REVIEW

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/server/AppiumServlet.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5802459&bid=18995379)

File: app/src/main/java/io/appium/uiautomator2/server/AppiumServlet.java [hunk 1]
```diff
@@ -281,7 +283,11 @@ private void addHandlerAttributesToRequest(IHttpRequest request, String mappedUr
 
         String id = getParameter(mappedUri, request.uri(), ":id");
         if (id != null) {
-            request.data().put(ELEMENT_ID_KEY, URLDecoder.decode(id));
+            try {
+                request.data().put(ELEMENT_ID_KEY, URLDecoder.decode(id, StandardCharsets.UTF_8.name()));
+            } catch (UnsupportedEncodingException e) {
+                throw new RuntimeException(e);
```
Comment (dpgraham) [inline app/src/main/java/io/appium/uiautomator2/server/AppiumServlet.java:289]: Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5802459)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5802459&bid=18995379)

File: app/src/main/java/io/appium/uiautomator2/server/AppiumServlet.java [hunk 2]
```diff
@@ -293,19 +295,19 @@ private void addHandlerAttributesToRequest(IHttpRequest request, String mappedUr
...
         for (int i = 0; i < currentSections.length; i++) {
-            if (configuredSections[i].contains(param)) {
-                return currentSections[i];
+            if (!configuredSections[i].contains(param)) {
+                continue;
+            }
+            try {
+                return URLDecoder.decode(currentSections[i], StandardCharsets.UTF_8.name());
+            } catch (UnsupportedEncodingException e) {
+                throw new RuntimeException(e);
```
Comment (dpgraham) [inline app/src/main/java/io/appium/uiautomator2/server/AppiumServlet.java:310]: Codacy found an issue: [Avoid throwing raw exception types.](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5802459)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/UiElementSnapshot.java  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5802459&bid=18995379)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/UiElementSnapshot.java  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5802459&bid=18995379)

**Comment by mykola-mokhnach** [comment]
@sshock I assume this change could also improve the perf of your tests if you use nested lookups

**Comment by sshock** [comment]
@mykola-mokhnach I don't think we are using any nested lookups, but happy to see more performance improvements and thanks for tagging me!

**Pull request merged**